### ISSUE TITLE
Support SSL for v0.8.3

### ISSUE THREAD

**Issue by joestein** [edenhill/librdkafka#154]
Kafka is adding SSL support => https://issues.apache.org/jira/browse/KAFKA-1477 and would be great to have this available in c++ libraries.

**Comment by thelamb** [comment]
The official SSL support for Kafka is late. For our usecase, SSL is a hard requirement so I've started looking into implementing SSL in rdkafka. There is a fork for the Kafka server available that supposedly implements SSL: https://github.com/relango/kafka/tree/kafka_security_0.8.2

I've tested this implementation and unfortunately it appears incomplete and unstable. The kafka server (./bin/kafka-server-start.sh config/server.properties) crashes after a few minutes because of 'Too many open files'. The consumer (./bin/kafka-simple-consumer-shell.sh --broker-list=localhost:9092 --secure --topic test --partition 0 --security.config.file config/client.security.properties) throws out of memory exceptions, even though there is a commit indicating that the author got the simple consumer to work.
Also the unit tests are failing miserably. 

So we're not confident in using this unofficial fork until the official support for SSL in Kafka is finished. 

Since we require SSL, I've started looking into using Kafka together with stunnel. I'll write a short overview of the solution here in case it is useful for someone else. Also any feedback is appreciated in case I've overlooked something. Once the official support for SSL is completed, I'll continue the work on implementing it in rdkafka. 

# Stunnel 
Website: https://www.stunnel.org/index.html
We're running everything on Ubuntu (sudo apt-get install stunnel4)

The idea is that we bind stunnel on the public IP address of the server, on the same port as Kafka (9092). The Kafka server will bind the same 9092 port on localhost. 
Stunnel will be configured to accept incoming connections on 9092 and forward them unencrypted to localhost:9092. 

## Server-side configuration 
    $ cat /etc/stunnel/stunnel.conf 
    client = no
    
    [kafka]
    accept = <your public IP>:9092
    connect = 127.0.0.1:9092
    cert = /etc/stunnel/stunnel.pem

The downside clearly is that your public IP has to be known. A hostname will not work here since the hostname will resolve to localhost.

The Kafka server needs to be configured to bind on localhost, rather than the default 0.0.0.0. 
In <Kafka-path>/config/server.properties we set:

    host.name=127.0.0.1

If you leave this default, the server will not be able to start because of 'Bind address already in use'.
The result of setting host.name to localhost is that the Broker will also advertise 127.0.0.1 to Producers and Consumers. This can be changed by setting advertised.host.name, but since we also have an stunnel at the Producer/Consumer side, we actually want the Producer/Consumer to connect to 127.0.0.1. 

## Client-side configuration
    $ cat /etc/stunnel/stunnel.conf 
    cert = /etc/stunnel/stunnel.pem
    client = yes
    
    [kafka]
    accept = 127.0.0.1:9092
    connect = sm-lap-7:9092

So an unencrypted connection by the Consumer/Producer on localhost:9092 will be forwarded encrypted to sm-lap-7:9092, which is the stunnel port on the server side. 

# Fin

Clearly this solution is not ideal. The Broker hostnames/IP addresses need to be in the stunnel configuration on the Consumer/Producer side, and each Broker will need to listen on a different port because the Consumer/Producers can only connect to localhost:<port>. Also each broker needs to have its public IP address in the stunnel configuration.

For us, this works fine for the time being because of the small number of Brokers (1 or 2), all with static IP addresses. I just wanted to share this work hoping that it is useful for anyone else. 

As I said before, I will work on the SSL implementation in rdkafka once the official SSL support (nears) completion. By that time we'll also have to support more brokers and consumers.

**Comment by edenhill** [comment]
Thanks for the detailed analysis.

The SSL support is up for discussion in [KIP-12](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=51809888)
Please have a look to see that they are on the right track, more eyes the better.

I do agree that there isn't any point in starting client side work before the KIP has settled.

I've refactored the communication code a bit in the dev branch that will make adding an SSL layer easier.


Regarding stunnel: It is quite a nice and reusable workaround for the time being.
What could we do in rdkafka to make stunnel integration easier?
I guess it boils down to address translation, if the broker reports address:port X:Y, we want to convert that to A:B, where A is probably localhost.

One alternative would be add a configurable resolve callback which will be called prior to each connect() attempt for each broker (either from the bootstrap broker list or from broker metadata),
returning an IP address (v4 or v6) as well as an alternate port.

This could be useful in other scenarios as well.

Something like:
````
rd_kafka_conf_set_resolve_cb (rd_kafka_conf_t *conf,
     int (*resolve_cb) (rd_kafka_t *rk, const char *host,
                                  some_way_to_portably_return_a_list_of_family_agnostic_sockaddrs,
                                 char *errstr, size_t errstr_size);
````

What do you think of that?



Another alternative would be to use SOCKS which allows the client (rdkafka) to specify the remote peer specifically.

**Comment by thelamb** [comment]
I will have a look at the KIP. 

> What could we do in rdkafka to make stunnel integration easier?
> I guess it boils down to address translation,

As far as I know, the address translation will only be necessary if the Broker needs to support both encrypted (through stunnel) and unencrypted (public) connections. In our scenario, we only want encrypted connections to the Broker:

    <Server>
     Stunnel: Listen on <PublicBrokerIP>:9092
     Broker: Listen on localhost:9092

    <Client>
     Stunnel: Listen on localhost:9092, forward encrypted to <PublicBrokerIP>:9092 

The Broker has host.name configured as `127.0.0.1`, so it is going to advertise `127.0.0.1:9092` to Consumers and Producers. Consumers/Producers will connect to this advertised `127.0.0.1:9092`, which is an stunnel instance on the Consumer/Producer that forwards the connection encrypted to `<PublicBrokerIP>:9092`. The Consumer/Producer does not need knowledge of this `<PublicBrokerIP>`. 

So I believe there is nothing we can do on rdkafka side to make this any easier. 

If the Broker needs to support both encrypted and unencrypted, then we do need some work on rdkafka side, because only rdkafka knows whether the connection to the Broker should be encrypted or not. I can foresee something like this:

    <Server>
     Stunnel: Listen on <PublicBrokerIP>:9093, forward unencrypted to localhost:9092
     Broker: Listen on 0.0.0.0:9092

    <Client>
     Stunnel: Listen on localhost:9092, forward encrypted to <PublicBrokerIP>:9093 
   
Now the Broker does not have `host.name=127.0.0.1`, since it needs to bind to `0.0.0.0:9092` to allow for public, unenctyped connections. So it is going to advertise `<PublicBrokerIP>:9092` to Consumers/Producers. Without any work on rdkafka side, we should see the following (undesired) behavior:

     <rdkafka> -> localhost:9092 -> <PublicBrokerIP>:9093  (first connection because of --broker-list localhost:9092)
                                 ^^ stunnel magic
     <broker> -> <rdkafka> : Leader = <PublicBrokerIP>:9092   (Broker advertises <PublicBrokerIP>)

     <rdkafka> -> <PublicBrokerIP>:9092  
               ^^ Unencrypted, not going through stunnel

So in this case, rdkafka would need to translate` <PublicBrokerIP>:9092` to `localhost:9092`.

I personally don't have a use for this, all of our connections to the Broker will be encrypted so having the Broker advertise `127.0.0.1:9092` works perfectly. 
Once the KIP is completed, the Broker should support both encrypted and unencrypted connections through this JIRA (https://issues.apache.org/jira/browse/KAFKA-1809) (Refactor brokers to allow listening on multiple ports and IPs). I assume that the Broker will advertise the same IP address to which the client actually connected. 

So I personally wouldn't spend time on implementing the address translation in rdkafka to support this, when it probably won't be needed any more once the official SSL implementation is finished

**Comment by edenhill** [comment]
@thelamb If you are still in need of SSL support please try out the `dev` branch and let me know how it works out for you.

**Comment by edenhill** [comment]
Instructions here: https://github.com/edenhill/librdkafka/wiki/Using-SSL-with-librdkafka

**Comment by thelamb** [comment]
That's pretty good timing @edenhill. We were about to make the stunnel 'solution' ready for production. I'll have a look at the latest dev and let you know how it goes.

**Comment by thelamb** [comment]
FYI, we've tested the SSL implementation and it seems to work well. Unfortunately for us we still have to use stunnel on the broker side because there are consumers (druid) that have no support for SSL yet. 

I'm happy though since not having stunnel on the producer where we use rdkafka really lowers the maintenance burden. I only had to hack rdkafka to ignore the 'advertised.host.name' parameter coming from the broker, since this has to be set to 127.0.0.1 for the stunnel to work (see above).
At the moment we only have 1 fixed broker so this is not really a problem. I hope that by the time we need multiple brokers, the official SSL support will land in Kafka and Druid :)

Thanks as always for your effort @edenhill
